Basic integration with the py-spy profiler

jarlsondre opened the pull request:
# Basic Integration with the py-spy Profiler

The goal of this PR is to add a basic level of integration with the py-spy profiler. This includes: 

- [x] Adding py-spy to the project dependencies
- [x] CLI functionality for aggregating some basic output
- [x] Launching py-spy in the SLURM script with support for multi-GPU training
- [x] CLI functionality for making a [Flamegraph](https://github.com/brendangregg/FlameGraph)

## Basic Data Aggregation

The basic data aggregation consists of finding all the final function calls in each stack and sorting them by number of samples, which is a proxy of how much time was spent in them. This is printed as a table and the idea is that you expect most of the time to be spent in the backward and forward calls from PyTorch. If you see something else at the top, then this might suggest a bottleneck. 

Example of table output (with and without aggregation):
```
❯ itwinai generate-py-spy-report --file profiling_data.txt
 name                      | path                       |   line | itwinai_function_name      | itwinai_function_path    |   itwinai_function_line | proportion (n)
---------------------------+----------------------------+--------+----------------------------+--------------------------+-------------------------+------------------
 _engine_run_backward      | torch/autograd/graph.py    |    769 | train_step                 | itwinai/torch/trainer.py |                     692 | 31.74% (908)
 _conv_forward             | torch/nn/modules/conv.py   |    454 | train_step                 | itwinai/torch/trainer.py |                     690 | 8.88% (254)
 _conv_forward             | torch/nn/modules/conv.py   |    454 | train_step                 | itwinai/torch/trainer.py |                     690 | 6.26% (179)
 _max_pool2d               | torch/nn/functional.py     |    796 | train_step                 | itwinai/torch/trainer.py |                     690 | 3.95% (113)
 dropout2d                 | torch/nn/functional.py     |   1387 | train_step                 | itwinai/torch/trainer.py |                     690 | 1.71% (49)
 append_to                 | mlflow/utils/file_utils.py |    505 | log                        | itwinai/loggers.py       |                     636 | 1.40% (40)
 _conv_forward             | torch/nn/modules/conv.py   |    454 | validation_step            | itwinai/torch/trainer.py |                     784 | 1.26% (36)
 _max_pool2d               | torch/nn/functional.py     |    796 | train_step                 | itwinai/torch/trainer.py |                     690 | 1.26% (36)
 _conv_forward             | torch/nn/modules/conv.py   |    454 | validation_step            | itwinai/torch/trainer.py |                     784 | 1.01% (29)
 _call_with_frames_removed | built-in                   |     -1 | exec_pipeline_with_compose | itwinai/cli.py           |                     460 | 0.87% (25)
❯ itwinai generate-py-spy-report --file profiling_data.txt --aggregate-leaf-paths
 name                      | path                       |   line | itwinai_function_name      | itwinai_function_path    |   itwinai_function_line | proportion (n)
---------------------------+----------------------------+--------+----------------------------+--------------------------+-------------------------+------------------
 _engine_run_backward      | torch/autograd/graph.py    |    769 | train_step                 | itwinai/torch/trainer.py |                     692 | 31.74% (908)
 _conv_forward             | torch/nn/modules/conv.py   |    454 | train_step                 | itwinai/torch/trainer.py |                     690 | 15.13% (433)
 _max_pool2d               | torch/nn/functional.py     |    796 | train_step                 | itwinai/torch/trainer.py |                     690 | 5.21% (149)
 _conv_forward             | torch/nn/modules/conv.py   |    454 | validation_step            | itwinai/torch/trainer.py |                     784 | 2.27% (65)
 append_to                 | mlflow/utils/file_utils.py |    505 | log                        | itwinai/loggers.py       |                     636 | 2.17% (62)
 list_all                  | mlflow/utils/file_utils.py |    146 | log                        | itwinai/loggers.py       |                     636 | 1.99% (57)
 _call_with_frames_removed | built-in                   |     -1 | exec_pipeline_with_compose | itwinai/cli.py           |                     460 | 1.82% (52)
 dropout2d                 | torch/nn/functional.py     |   1387 | train_step                 | itwinai/torch/trainer.py |                     690 | 1.71% (49)
 get_single_data           | yaml/constructor.py        |     49 | log                        | itwinai/loggers.py       |                     636 | 1.22% (35)
 isdir                     | <frozen genericpath>       |     42 | log                        | itwinai/loggers.py       |                     636 | 0.98% (28)
```

matbun commented on src/itwinai/flamegraph.pl:
if you haven't yet, check how files from external projects should be embedded into our repo. For instance, I think we need to:

1. Check that their license does not clash with ours
2. Add a header to this file, giving credits to the authors. For instance, copying the LICENSE statement on their repo.
3. Update our LICENSE file, or similar, explaining that this file is coming from somewhere else

matbun commented on src/itwinai/slurm/utils.py:
```diff
@@ -67,6 +67,7 @@ def get_slurm_job_parser() -> ArgumentParser:
     default_training_command = None
     default_python_venv = ".venv"
     default_scalability_nodes = "1,2,4,8"
+    default_profiling_rate = 10
```
Maybe add a small doc in a comment above this line to explain what this is for?

lineick commented:
Printing multiple lines for identical function calls is not a good idea if one cannot relate them back to the call stack in the file.

I would suggest aggregating identical function calls for the printed overview. Otherwise if there are identifiers per call stack, one can also add those as a column to the overview (but that may be a bit confusing too).

jarlsondre commented:
> Printing multiple lines for identical function calls is not a good idea if one cannot relate them back to the call stack in the file.
> 
> I would suggest aggregating identical function calls for the printed overview. Otherwise if there are identifiers per call stack, one can also add those as a column to the overview (but that may be a bit confusing too).

Added this feature now, also updated the screenshot at the top
